src/rshim.c

```diff
@@ -1746,7 +1787,17 @@ static void rshim_work_handler(rshim_backend_t *bd)
       bd->timer = rshim_timer_ticks + 100;
   }
 
+  if (bd->has_locked_work) {
+    bd->has_locked_work = 0;
+    rshim_update_locked_mode(bd);
+  }
   pthread_mutex_unlock(&bd->mutex);
+
+  /* Ownership transfer needs to run without holding bd->mutex. */
+  if (bd->has_osp_work) {
+    bd->has_osp_work = 0;
+    handle_ownership_transfer(bd);
+  }
```

Comment:
Need to double check. All rshim read/write should be protected by bd->mutex

Reply:
handle_ownership_transfer() calls rshim_set_drop_mode(), but previous implementation of rshim_set_drop_mode() holds the bd->mutex. That's why it cannot be inside the lock previously. This afternoon's drop_mode rewrite has removed the locking in rshim_set_drop_mode(). So It's possible to put this section back inside lock again. Will change.